tests/test_ensemble_tools.py

```diff
@@ -0,0 +1,163 @@
+        msg = f"Failed to find PARAMETERS in the \
+        CONDITIONS node of {conditions_file_name}"
+        raise ValueError(msg)
+
+    for element in parameters.iter("P"):
+        match = re.match(
+            r"\s*(?P<key>\w+)\s*=\s*(?P<value>-?\d*(\.?\d)+)\s*", element.text
+        )
+        if match is None:
+            msg = f"Parameter definition of unexpected format: {element.text}"
+            raise ValueError(msg)
+        key = match.group("key")
+        value = float(match.group("value"))
+        if key == parameter_name:
+            return expected_value == pytest.approx(value)  # Because of rounding errors
+
+    # If we make it this far then we haven't found the parameter we were looking for,
+    # which is surely a problem in its own right
+    msg = f"Could not find the parameter {parameter_name} in {conditions_file_name}"
+    raise ValueError(msg)
+
+
+def test_create_dir_tree(tmpdir: local) -> None:
+    """Test the create_dir_tree method of ensemble_tools."""
+    test_sweep_dir = Path(tmpdir) / "test_sweep"
+    n_dirs = 5
+    destructive = True
+    copy_dir = Path("config_files") / "two_stream"
+    edit_file = "conditions.xml"
+    parameter_to_scan = "particle_initial_velocity"
+    scan_range = (0.2, 5.0)
+    outdir_prefix = "asd"
+
+    create_dir_tree(
+        sweep_path=test_sweep_dir,
+        n_dirs=n_dirs,
+        destructive=destructive,
+        copy_dir=copy_dir,
+        edit_file=edit_file,
+        parameter_to_scan=parameter_to_scan,
+        scan_range=scan_range,
+        outdir_prefix=outdir_prefix,
+    )
+
+    assert (test_sweep_dir / "SWEEP").is_dir()
+    assert len(list((test_sweep_dir / "SWEEP").iterdir())) == n_dirs
+    for i in range(n_dirs):
+        cond_file = test_sweep_dir / "SWEEP" / f"asd{i}" / "conditions.xml"
+        assert cond_file.is_file()
+        para_value = scan_range[0] + (i / (n_dirs - 1)) * (
+            scan_range[1] - scan_range[0]
+        )
+        assert _check_parameter_in_conditions(cond_file, parameter_to_scan, para_value)
+    # Should raise an exception when trying to edit the mesh file
+    edit_file = "mesh.xml"
+    with pytest.raises(ValueError, match=r".* a CONDITIONS node.*"):
+        create_dir_tree(
+            sweep_path=test_sweep_dir,
+            n_dirs=n_dirs,
+            destructive=destructive,
+            copy_dir=copy_dir,
+            edit_file=edit_file,
+            parameter_to_scan=parameter_to_scan,
+            scan_range=scan_range,
+            outdir_prefix=outdir_prefix,
+        )
+
+
+def test_create_dict_sweep(tmpdir: local) -> None:
+    """Test the create_dict_sweep method of ensemble_tools."""
+    sweep_path = Path(tmpdir) / "test"
+    n_divs = 5
+    destructive = True
+    copy_dir = Path("config_files") / "two_stream"
+    edit_file = "conditions.xml"
+    parameter_dict = {
+        "particle_initial_velocity": (0.1, 2.5),
+        "particle_charge_density": (102.0, 108.0),
+    }
```

Comment:
As in earlier comment for `test_create_dir_tree` I would parameterize `n_divs` here to ensure we test edge case `n_divs = 1` as well as typical case, and probably `copy_dir` / `parameter_dict` and `destructive`.

Reply:
Similarly, I've parametrized those mentioned variables, I can add further test cases if there are any other obvious edge cases you can think of.